File: android/app/src/main/java/com/lasercars_mobile/DrawingThread.java
package com.lasercars_mobile;

import android.graphics.Canvas;
import android.util.Log;
import android.view.SurfaceHolder;

public class DrawingThread extends Thread {
    private SurfaceHolder surfaceHolder;
    private SkiaCanvas skiaCanvas;
    private boolean isRunning = false;
    private long previousTime;
    private final int fps = 60;

    public DrawingThread(SurfaceHolder surfaceHolder, SkiaCanvas skiaCanvas) {
        this.surfaceHolder = surfaceHolder;
        this.skiaCanvas = skiaCanvas;

        previousTime = System.currentTimeMillis();
    }

    public void setRunning(boolean run) {
        isRunning = run;
    }

    @Override
    public void run() {
        Canvas canvas;

        while (isRunning) {

            long currentTimeMillis = System.currentTimeMillis();
            long elapsedTimeMs = currentTimeMillis - previousTime;
            long sleepTimeMs = (long) (1000f / fps - elapsedTimeMs);

            canvas = null;
            try {

                canvas = surfaceHolder.lockCanvas();

                if (canvas == null) {
                    Thread.sleep(1);
                    continue;
                } else if (sleepTimeMs > 0) {
                    Thread.sleep(sleepTimeMs);
                }

                synchronized (surfaceHolder) {
                    currentTimeMillis = System.currentTimeMillis();
                    elapsedTimeMs = currentTimeMillis - previousTime;
                    double percentageOfFrame = (double) elapsedTimeMs / (1000f / fps);
                    skiaCanvas.update(percentageOfFrame, canvas);
                }
            } catch (Exception e) {
                e.printStackTrace();
            } finally {
                if (canvas != null) {
                    surfaceHolder.unlockCanvasAndPost(canvas);
                    previousTime = System.currentTimeMillis();
                }
            }
        }
    }
}

File: android/app/src/main/java/com/lasercars_mobile/DrawingThread.kt
package com.lasercars_mobile

import android.graphics.Canvas
import android.view.SurfaceHolder

class DrawingThread(private var surfaceHolder: SurfaceHolder, private var skiaCanvas: SkiaCanvas) :
    Thread() {
    private var isRunning = false
    private var previousTime: Long = System.currentTimeMillis()
    private val fps = 60

    fun setRunning(run: Boolean) {
        isRunning = run
    }

    override fun run() {
        var canvas: Canvas?
        var runOnce = false

        while (isRunning || !runOnce) {
            var elapsedTimeMs: Long
            val sleepTimeMs: Long

            canvas = null
            try {
                canvas = surfaceHolder.lockCanvas()

                if (canvas != null) {
                    synchronized(surfaceHolder) {
                        elapsedTimeMs = System.currentTimeMillis() - previousTime
                        val percentageOfFrame = elapsedTimeMs.toDouble() / (1000f / fps)
                        previousTime = System.currentTimeMillis()
                        skiaCanvas.update(percentageOfFrame, canvas)
                        runOnce = true
                    }
                }
            } catch (e: Exception) {
                e.printStackTrace()
            } finally {
                if (canvas != null) {
                    surfaceHolder.unlockCanvasAndPost(canvas)
                    elapsedTimeMs = System.currentTimeMillis() - previousTime
                    sleepTimeMs = (1000f / fps - elapsedTimeMs).toLong()

                    if (sleepTimeMs > 0) {
                        sleep(sleepTimeMs)
                    }
                } else {
                    sleep(1)
                }
            }
        }
    }
}